bentoml/_internal/bento/bento.py

```diff
@@ -288,8 +315,77 @@ def export(self, path: str):
 
-    def push(self):
-        pass
+    @inject
+    def push(self, *, force: bool = False, model_store: "ModelStore" = Provide[BentoMLContainer.model_store]):
+        yatai_cli = get_current_yatai_cli()
+        name = self.tag.name
+        version = self.tag.version
+        info = self.info
+        model_names = info.models
+        with ThreadPoolExecutor(max_workers=max(len(model_names), 1)) as executor:
+            futures = executor.map(lambda x: x.push(force=force), (model_store.get(name) for name in model_names))
+            list(futures)
+        with yaspin(text=f'Fetching bento {name}'):
+            bento_obj = yatai_cli.get_bento(bento_name=name)
+        if not bento_obj:
+            bento_obj = yatai_cli.create_bento(req=CreateBentoSchema(name=name, description=''))
+        with yaspin(text=f'Fetching bento version {version}'):
+            bento_version_obj = yatai_cli.get_bento_version(bento_name=name, version=version)
+        if not force and bento_version_obj and bento_version_obj.upload_status == BentoVersionUploadStatus.SUCCESS:
+            logger.info(f"Bento {self.tag} already exists in yatai, skipping.")
+            return
+        if not bento_version_obj:
+            labels: t.List[LabelItemSchema] = [LabelItemSchema(key=key, value=value) for key, value in info.labels.items()]
+            apis: t.Dict[str, BentoVersionApiSchema] = {}
+            yatai_cli.create_bento_version(bento_name=bento_obj.name, req=CreateBentoVersionSchema(
+                description='',
+                version=version,
+                build_at=info.creation_time,
+                manifest=BentoVersionManifestSchema(
+                    service=info.service,
+                    bentoml_version=info.bentoml_version,
+                    apis=apis,
+                    models=info.models,
+                ),
+                labels=labels,
+            ))
+        bento_version_obj = yatai_cli.presign_bento_version_s3_upload_url(bento_name=bento_obj.name, version=version)
+        with io.BytesIO() as tar_io:
+            bento_dir_path = self.path
+            with yaspin(text=f'Taring bento {self.tag}'):
+                with tarfile.open(fileobj=tar_io, mode='w:gz') as tar:
+                    def filter_(tar_info: tarfile.TarInfo) -> t.Optional[tarfile.TarInfo]:
+                        if tar_info.path == './models' or tar_info.path.startswith('./models/'):
+                            return None
+                        return tar_info
+                    tar.add(bento_dir_path, arcname='./', filter=filter_)
+            tar_io.seek(0, 0)
+            with yaspin(text=f'Starting upload bento {self.tag}'):
+                yatai_cli.start_upload_bento_version(bento_name=bento_obj.name, version=version)
+            file_size = tar_io.getbuffer().nbytes
+            with tqdm(desc=f'Uploading bento {self.tag}', total=file_size, unit='B', unit_scale=True, unit_divisor=1024) as progress_bar:
```

Comment:
should we just use the progress bar from Rich? https://github.com/willmcgugan/rich#rich-library to avoid adding another dependency?

Reply:
Yes, I'm trying to use rich's progressbar, but the API of rich's progressbar is really stupid compared to tqdm